### PR TITLE
WT-7666 Add assertion to check whether duplicate history store inserts are modifies

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
We're currently having issues with duplicate insertions into the history store. This is not harmful in the case of full updates but could cause invalid values to be read for modifies.

While we're trying to solve this, I want to gain an understanding of whether this can only happen with full updates so I'm adding an assertion to distinguish between these two cases.

### CODE REVIEW

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -150,16 +150,28 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
+             * 2. it came from a different transaction
+             * 3. it came from the same transaction but with a different timestamp
              */
-            if (cmp == 0)
-                WT_ASSERT(session,
-                  __wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw) ||
-                    tw->start_txn == WT_TXN_NONE ||
-                    tw->start_txn != hs_cbt->upd_value->tw.start_txn ||
-                    tw->start_ts != hs_cbt->upd_value->tw.start_ts);
+            if (cmp == 0) {
```
Comment (quchenhao) [inline src/history/hs_rec.c:159]: Is it really needed? If we hit the assert, we can read the type from the core.

Comment (tetsuo-cpp) [reply src/history/hs_rec.c:159]: I'd say it's useful.

Yeah, we can inspect from the core. But there are going to be multiple BFGs and they're going to be grouped under the same BF regardless of the update type. We're not going to know if one of those BFGs was a modify unless we spawn a host and inspect every single core which isn't feasible.

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -150,16 +150,28 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
+                if (!__wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw) &&
```
Comment (quchenhao) [inline src/history/hs_rec.c:160]: Can we wrap this in HAVE_DIAGNOSTIC so it is not checked in production build?

Comment (tetsuo-cpp) [reply src/history/hs_rec.c:160]: Yep. This check is already marked in a `HAVE_DIAGNOSTIC`.

File: src/history/hs_rec.c (hunk 1 above)
Comment (quchenhao) [reply src/history/hs_rec.c:159]: oh, right

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by sulabhM — APPROVED** [review]
LGTM

**Pull request merged**